### DISCUSSION TITLE
Help with user fork build of ORFS

### DISCUSSION THREAD

**Discussion by gudeh** [The-OpenROAD-Project/OpenROAD-flow-scripts#1572] [Q&A] [answered]
Hey guys, I have been working with my own fork of ORFS for about a year:

I have always been able to keep up to date with latest changes of ORFS on my fork. Although recently I got into some compilation errors which I could not handle by myself. 

If I leave .gitmodules file as is, for some reason I get the following errors, right at the beginning of "./build_openroad.sh":
```
[INFO FLW-0002] Updating git submodules.
Submodule 'tools/LSOracle' (https://github.com/The-OpenROAD-Project/LSOracle.git) registered for path 'tools/LSOracle'
Submodule 'tools/OpenROAD' (https://github.com/gudeh/OpenROAD.git) registered for path 'tools/OpenROAD'
Submodule 'tools/yosys' (https://github.com/The-OpenROAD-Project/yosys.git) registered for path 'tools/yosys'
Cloning into '/home/gudeh/Desktop/OpenROAD-flow-scripts/tools/LSOracle'...
Cloning into '/home/gudeh/Desktop/OpenROAD-flow-scripts/tools/OpenROAD'...
remote: Support for password authentication was removed on August 13, 2021.
remote: Please see https://docs.github.com/en/get-started/getting-started-with-git/about-remote-repositories#cloning-with-https-urls for information on currently recommended modes of authentication.
fatal: Authentication failed for 'https://github.com/gudeh/OpenROAD.git/'
fatal: clone of 'https://github.com/gudeh/OpenROAD.git' into submodule path '/home/gudeh/Desktop/OpenROAD-flow-scripts/tools/OpenROAD' failed
Failed to clone 'tools/OpenROAD'. Retry scheduled
Cloning into '/home/gudeh/Desktop/OpenROAD-flow-scripts/tools/yosys'...
Cloning into '/home/gudeh/Desktop/OpenROAD-flow-scripts/tools/OpenROAD'...
remote: Support for password authentication was removed on August 13, 2021.
remote: Please see https://docs.github.com/en/get-started/getting-started-with-git/about-remote-repositories#cloning-with-https-urls for information on currently recommended modes of authentication.
fatal: Authentication failed for 'https://github.com/gudeh/OpenROAD.git/'
fatal: clone of 'https://github.com/gudeh/OpenROAD.git' into submodule path '/home/gudeh/Desktop/OpenROAD-flow-scripts/tools/OpenROAD' failed
Failed to clone 'tools/OpenROAD' a second time, aborting
```
Notice how it sets up a git url with my user, which it did not use to happen before. For this reason I modified .gitmodules with the complete path and got another error.

If I change .gitmodules like the following, the build goes on until it provides the following error (I am skipping installing of Yosys because of EQY issue):

.gitmodule:
```
[submodule "tools/yosys"]
	path = tools/yosys
	url = ../../The-OpenROAD-Project/yosys.git

[submodule "tools/OpenROAD"]
path = tools/OpenROAD
url = https://github.com/The-OpenROAD-Project/OpenROAD.git

[submodule "tools/LSOracle"]
path = tools/LSOracle
url = https://github.com/The-OpenROAD-Project/LSOracle.git
```
Build error:
```
[ 88%] Building CXX object src/drt/CMakeFiles/drt.dir/src/distributed/frArchive.cpp.o
[ 88%] Building CXX object src/drt/CMakeFiles/drt.dir/src/distributed/drUpdate.cpp.o
[ 88%] Building CXX object src/drt/CMakeFiles/drt.dir/src/distributed/paUpdate.cpp.o
[ 88%] Building CXX object src/drt/CMakeFiles/drt.dir/src/TritonRoute.cpp.o
[ 89%] Building CXX object src/drt/CMakeFiles/drt.dir/src/MakeTritonRoute.cpp.o
[ 89%] Building CXX object src/drt/CMakeFiles/drt.dir/src/frBaseTypes.cpp.o
[ 89%] Building CXX object src/drt/CMakeFiles/drt.dir/src/DesignCallBack.cpp.o
[ 89%] Linking CXX static library drt.a
[ 89%] Built target drt
gmake: *** [Makefile:146: all] Error 2
```

If use the original ORFS it compiles successfully and runs with manual installed EQY. 

I also have a question. Can I install multiple versions of ORFS on the same system? Are all the tools contained to the project folder, or does ORFS installs binaries on the system?

**Comment by gudeh** [comment]
[build_openroad-gitmoduleEditted.log](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/files/13032021/build_openroad-gitmoduleEditted.log)

**Comment by maliberty** [comment]
I'm guessing this relates to https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/1370 .  If you locally undo that change does it solve your problem?

@vvbandeira any ideas?

**Reply by gudeh** [reply]
I tried to undo the modifications on .gitmodules, retry "build_openroad.sh --local" (without clean) and the build error shows up again in another percentage progression, although with the same error text:

```
[ 94%] Building CXX object src/drt/CMakeFiles/drt.dir/src/distributed/drUpdate.cpp.o
[ 94%] Building CXX object src/drt/CMakeFiles/drt.dir/src/distributed/paUpdate.cpp.o
[ 94%] Building CXX object src/drt/CMakeFiles/drt.dir/src/TritonRoute.cpp.o
[ 95%] Building CXX object src/drt/CMakeFiles/drt.dir/src/MakeTritonRoute.cpp.o
[ 95%] Building CXX object src/drt/CMakeFiles/drt.dir/src/frBaseTypes.cpp.o
[ 95%] Building CXX object src/drt/CMakeFiles/drt.dir/src/DesignCallBack.cpp.o
[ 95%] Linking CXX static library drt.a
[ 95%] Built target drt
gmake: *** [Makefile:146: all] Error 2
```

**Reply by gudeh** [reply]
[build_openroad.log](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/files/13032964/build_openroad.log)

**Reply by maliberty** [reply]
The error
```
/home/gudeh/Desktop/OpenROAD-flow-scripts/tools/OpenROAD/src/pdn/src/sroute.cpp:507:14: error: ‘cout’ is not a member of ‘std’
  507 |         std::cout << "xmax is " << pdn_wire->xMax() << std::endl;
      |              ^~~~
```
has been fixed.  Please update and try again.

**Comment by vvbandeira** [comment]
@gudeh 
- Have you tried to clone the repo again?
- You can have as many ORFS clones as you want in your system, the `build_openroad.sh` just builds the binaries. They are not installed in your system by default.
- Which version of `git` and which OS are you using?

**Reply by gudeh** [reply]
@vvbandeira 

> Have you tried to clone the repo again?

Yes. Actually, I was only  able to build on my fork by doing "git reset --hard <hash>" to a commit of mine from april: https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/commit/db55b0891336d7790093a704cee99816966e97e5
I guess that I was able to build with merges between April and now because I did not go for a clean build.

> You can have as many ORFS clones as you want in your system, the build_openroad.sh just builds the binaries. They are not installed in your system by default.

This is great under my situation. 

>Which version of git and which OS are you using?

```git version 2.34.1```
```
gudeh@snowflake: $ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 22.04.3 LTS
Release:	22.04
Codename:	jammy
```

**Reply by gudeh** [reply]
I wonder if everyone's forks are also having this build issue.

**Reply by vvbandeira** [reply, marked as answer]
Looking at your user's repos, it does seem that you don't have a fork of OR.
Could you try to fork that as well? 
https://github.com/The-OpenROAD-Project/OpenROAD

**Reply by gudeh** [reply]
Indeed that solved the problem! No need to modify .gitmodules!
